azure/services/subnets/spec.go

```diff
@@ -38,112 +40,77 @@ type SubnetSpec struct {
 	}
 
-	if !s.IsVNetManaged {
-		// TODO: change this to terminal error once we add support for handling them
-		return nil, errors.Errorf("custom vnet was provided but subnet %s is missing", s.Name)
+	subnet.Spec = asonetworkv1.VirtualNetworks_Subnet_Spec{
+		AzureName: s.Name,
+		Owner: &genruntime.KnownResourceReference{
+			Name: s.VNetName,
+		},
+		AddressPrefixes: s.CIDRs,
 	}
-	subnetProperties := armnetwork.SubnetPropertiesFormat{
-		AddressPrefixes: azure.PtrSlice(&s.CIDRs),
-	}
-
 	// workaround needed to avoid SubscriptionNotRegisteredForFeature for feature Microsoft.Network/AllowMultipleAddressPrefixesOnSubnet.
 	if len(s.CIDRs) == 1 {
-		subnetProperties = armnetwork.SubnetPropertiesFormat{
-			AddressPrefix: &s.CIDRs[0],
-		}
+		subnet.Spec.AddressPrefix = &s.CIDRs[0]
 	}
 
 	if s.RouteTableName != "" {
-		subnetProperties.RouteTable = &armnetwork.RouteTable{
-			ID: ptr.To(azure.RouteTableID(s.SubscriptionID, s.ResourceGroup, s.RouteTableName)),
+		subnet.Spec.RouteTable = &asonetworkv1.RouteTableSpec_VirtualNetworks_Subnet_SubResourceEmbedded{
+			Reference: &genruntime.ResourceReference{
+				ARMID: azure.RouteTableID(s.SubscriptionID, s.VNetResourceGroup, s.RouteTableName),
+			},
 		}
 	}
 
 	if s.NatGatewayName != "" {
-		subnetProperties.NatGateway = &armnetwork.SubResource{
-			ID: ptr.To(azure.NatGatewayID(s.SubscriptionID, s.ResourceGroup, s.NatGatewayName)),
+		subnet.Spec.NatGateway = &asonetworkv1.SubResource{
+			Reference: &genruntime.ResourceReference{
+				ARMID: azure.NatGatewayID(s.SubscriptionID, s.ResourceGroup, s.NatGatewayName),
+			},
 		}
 	}
 
 	if s.SecurityGroupName != "" {
-		subnetProperties.NetworkSecurityGroup = &armnetwork.SecurityGroup{
-			ID: ptr.To(azure.SecurityGroupID(s.SubscriptionID, s.ResourceGroup, s.SecurityGroupName)),
+		subnet.Spec.NetworkSecurityGroup = &asonetworkv1.NetworkSecurityGroupSpec_VirtualNetworks_Subnet_SubResourceEmbedded{
+			Reference: &genruntime.ResourceReference{
+				ARMID: azure.SecurityGroupID(s.SubscriptionID, s.VNetResourceGroup, s.SecurityGroupName),
+			},
 		}
 	}
 
-	serviceEndpoints := make([]armnetwork.ServiceEndpointPropertiesFormat, 0, len(s.ServiceEndpoints))
+	//nolint:prealloc // pre-allocating this slice isn't going to make any meaningful performance difference
+	// and makes it harder to keep this value nil when s.ServiceEndpoints is empty as is necessary.
+	var serviceEndpoints []asonetworkv1.ServiceEndpointPropertiesFormat
 	for _, se := range s.ServiceEndpoints {
-		se := se
-		serviceEndpoints = append(serviceEndpoints, armnetwork.ServiceEndpointPropertiesFormat{Service: ptr.To(se.Service), Locations: azure.PtrSlice(&se.Locations)})
+		serviceEndpoints = append(serviceEndpoints, asonetworkv1.ServiceEndpointPropertiesFormat{Service: ptr.To(se.Service), Locations: se.Locations})
 	}
-	subnetProperties.ServiceEndpoints = azure.PtrSlice(&serviceEndpoints)
+	subnet.Spec.ServiceEndpoints = serviceEndpoints
```

Comment:
> makes it harder to keep this value nil when s.ServiceEndpoints is empty as is necessary

We could have something like
```suggestion
if len(s.ServiceEndpoints) != 0 {
	serviceEndpoints := make([]asonetworkv1.ServiceEndpointPropertiesFormat, 0, len(s.ServiceEndpoints))
	for...
	.
	.
	subnet.Spec.ServiceEndpoints = serviceEndpoints
}
```

But happy with current implementation as `serviceEndpoints`'s scope is limited to this function.

Reply:
Right, but that additional necessary syntax doesn't outweigh the potential performance improvement IMO.